From: Aram Sargsyan <aram@isc.org>
Date: Sun, 13 Mar 2022 13:47:16 +0000
Subject: Fix dig error when trying the next server after a TCP connection
 failure

When encountering a TCP connection error while trying to initiate a
connection to a server, dig erroneously cancels the lookup even when
there are other server(s) to try, which results in an assertion failure.

Cancel the lookup only when there are no more queries left in the
lookup's queries list (i.e. `next` is NULL).

Origin: upstream, https://gitlab.isc.org/isc-projects/bind9/-/commit/0fb4fc1897c4098df08b10e180db843be172d42d
Bug: https://gitlab.isc.org/isc-projects/bind9/-/issues/3205
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/bind9/+bug/1964686
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/bind9/+bug/1964400

---
 bin/dig/dighost.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/bin/dig/dighost.c b/bin/dig/dighost.c
index 8d142ed..ec5190e 100644
--- a/bin/dig/dighost.c
+++ b/bin/dig/dighost.c
@@ -3273,6 +3273,9 @@ tcp_connected(isc_nmhandle_t *handle, isc_result_t eresult, void *arg) {
 		}
 
 		query_detach(&query);
+		if (next == NULL) {
+			cancel_lookup(l);
+		}
 		lookup_detach(&l);
 
 		if (next != NULL) {
